packages/nyc-test-coverage/src/interceptors.ts

```diff
@@ -0,0 +1,59 @@
+import {
+  proxySinks,
+  SignalInput,
+  WorkflowExecuteInput,
+  WorkflowInterceptorsFactory,
+  WorkflowInboundCallsInterceptor,
+  QueryInput
+} from '@temporalio/workflow';
+import { CoverageSinks } from './sinks';
+
+const { coverage } = proxySinks<CoverageSinks>();
+
+class CoverageInterceptor implements WorkflowInboundCallsInterceptor {
+  public async handleQuery(input: QueryInput, next: any): Promise<unknown> {
+    const ret = await next(input);
+
+    // @ts-ignore
+    const globalCoverage = global.__coverage__;
+    coverage.merge(JSON.parse(JSON.stringify(globalCoverage)));
+    clearCoverage(globalCoverage);
+
+    return ret;
+  }
+
+  public async handleSignal(input: SignalInput, next: any): Promise<void> {
+    const ret = await next(input);
+
+    // @ts-ignore
+    const globalCoverage = global.__coverage__;
+    coverage.merge(JSON.parse(JSON.stringify(globalCoverage)));
+    clearCoverage(globalCoverage);
+
+    return ret;
+  }
+
+  public async execute(input: WorkflowExecuteInput, next: any): Promise<unknown> {
+    const ret = await next(input);
+
+    // @ts-ignore
+    const globalCoverage = global.__coverage__;
+    coverage.merge(JSON.parse(JSON.stringify(globalCoverage)));
+    clearCoverage(globalCoverage);
+
+    return ret;
+  }
+}
+
+// Export the interceptors
+export const interceptors: WorkflowInterceptorsFactory = () => ({
+  inbound: [new CoverageInterceptor()],
+});
+
+function clearCoverage(coverage: any): void {
```

Comment:
Since we merge coverage on every query, signal, etc. we need to make sure we're not double counting executions. Think of it this way:

1. A signal comes in. nyc increments the signal handler's execution count to 1. We merge the workflow's current coverage state. 
2. Another instance of the same signal comes in. nyc increments the signal handler's execution count to 2. We merge the workflow's current coverage state, which means the coverage map that stores all the Workflows' coverage data thinks the signal handler executed 3 times (1 + 2), instead of 2 times.

Reply:
I figured that out already but didn't have a chance to reply.
Thanks for the clarification.

Reply:
Aren't there type definitions for global.__coverage__ that we can use?